tests/test_grouping.py

```diff
@@ -0,0 +1,341 @@
+        """Verifies if all vertices of the same colour are not connected."""
+
+        for colour in colouring.keys():
+
+            grouping = colouring[colour]
+            size_grouping = len(grouping)
+
+            for i in range(size_grouping):
+                for j in range(i + 1, size_grouping):
+                    vert_i = grouping[i][0]
+                    vert_j = grouping[j][0]
+
+                    if adjacency_matrix[vert_i][vert_j] == 1:
+                        return False
+
+        return True
+
+    def test_graph_colouring(self):
+        """Verify that random unweighted, undirected graph's colour is a valid solution."""
+
+        n_qubits = 8
+        adjacency_matrix = np.random.randint(2, size=(n_qubits, n_qubits))
+        np.fill_diagonal(adjacency_matrix, 0)
+
+        for i in range(n_qubits):
+            for j in range(i + 1, n_qubits):
+                adjacency_matrix[j, i] = adjacency_matrix[i, j]
+
+        lf_colouring = largest_first(np.asarray([list(range(n_qubits))]), adjacency_matrix)
+        rlf_colouring = recursive_largest_first(
+            np.asarray([list(range(n_qubits))]), adjacency_matrix
+        )
+
+        assert self.verify_graph_colour_solution(adjacency_matrix, lf_colouring)
+        assert self.verify_graph_colour_solution(adjacency_matrix, rlf_colouring)
+
+    def test_trivial_graph_colouring(self):
+        """Tests validity of graph colouring solution for a graph with no edges."""
+
+        n_qubits = 8
+        adjacency_matrix = np.zeros((n_qubits, n_qubits))
+
+        lf_colouring = largest_first(np.asarray([list(range(n_qubits))]), adjacency_matrix)
+        rlf_colouring = recursive_largest_first(
+            np.asarray([list(range(n_qubits))]), adjacency_matrix
+        )
+
+        assert self.verify_graph_colour_solution(adjacency_matrix, lf_colouring)
+        assert self.verify_graph_colour_solution(adjacency_matrix, rlf_colouring)
+
+
+class TestOptimizeMeasurements:
+    """Tests for the `optimize_measurements` function."""
+
+    def test_optimize_measurements_qwc_generic_case(self):
+        """Generic test case without coefficients."""
+
+        observables = [PauliY(0), PauliX(0) @ PauliX(1), PauliZ(1)]
+
+        diagonalized_groupings_sol = [
+            [PauliZ(wires=[0]) @ PauliZ(wires=[1])],
+            [PauliZ(wires=[0]), PauliZ(wires=[1])],
+        ]
+
+        diagonalized_groupings = optimize_measurements(
+            observables, grouping="qwc", colouring_method="rlf"
+        )[1]
+
+        assert len(diagonalized_groupings) == len(diagonalized_groupings_sol)
+
+        for i in range(len(diagonalized_groupings_sol)):
+            assert len(diagonalized_groupings[i]) == len(diagonalized_groupings_sol[i])
+            for j in range(len(diagonalized_groupings_sol[i])):
+                assert are_identical_pauli_words(
+                    diagonalized_groupings[i][j], diagonalized_groupings_sol[i][j]
+                )
```

Comment:
Find this part to be a bit hard to follow. Might be worth seeing if readability can be boosted by creating boolean values using `all`/`any` built-ins, assigning objects some descriptive names and later having `assert` statements for only a few booleans